docs/adrs/005-jobs.md

```diff
@@ -0,0 +1,285 @@
+# ADR 005: What's a Job?
+
+We want to be able to specify builds in Roc (I mean, it's right there in the name!)
+
+To do that, we need to have a good idea about the kinds of data we're working with and how they relate to one another.
+
+## Jobs
+
+A **job** defines the basic unit of work in an rbt build.
+It specifies the set of inputs that trigger a rebuild and, in most cases, produces new outputs.
+
+### Commands
+
+Commands are the only truly required part of a job.
+Without them, nothing else makes sense!
+A simple job might look like this:
+
+```roc
+hello : Job
+hello =
+    job { command: exec echo [ "Hello, World!" ] }
+```
+
+The first argument `echo` here is a `Tool` (see below for how to define those.)
+The second argument is an array of arguments which will be passed to the tool when it executes.
+
+Of course, since Roc is a full programming language, you can define your own helpers easily:
+
+```roc
+execShellScript : String -> Command
+execShellScript = \script ->
+    exec sh [ "-c", script ]
+```
+
+(n.b. I'll be using the `execShellScript` helper throughout this document to make the examples a little bit more focused!)
+
+- [ ] TBD whether it makes sense for rbt to export a small standard library including things like `sh` or `bash`.
+      As of this writing, I'm learning "no".
+
+- [ ] Is "exec" or "run" a better name?
+
+### Environment
+
+Of course, commands often need environment variables to work properly, so you can specify those:
+
+```roc
+hello : Job
+hello =
+  job
+      {
+          command: execShellScript "echo $GREETING",
+          environment: {:
+            "GREETING" => "Hello, World!",
+          :},
+      }
+```
+
+### Inputs
+
+Jobs must specify their complete set of inputs, and a job will be rebuilt any time any of its inputs change.
+Jobs will not be able to see any files outside the inputs they specify (see [ADR #1](./001-job-isolation-targets.md).)
+
+There are two types of inputs:
+
+- Files from the filesystem (*only* files, not globs or directories.
+  However, a future ADR will describe a way to automatically discover files.)
+- The outputs of other jobs.
+
+You might specify a job with file inputs like this:
+
+```roc
+app : Job
+app =
+    job
+        {
+           command: exec elm [ "make", "--output=app.js", "src/Main.elm" ],
+           inputFiles: [ "elm.json", "src/Main.elm" ],
+           outputFiles: [ "app.js" ],
+        }
+```
+
+And one which builds on another job like this:
```

Comment:
Might be worth noting that the reason `inputJobs` is benefitical compared to something like this...

```elm
inputFiles: List.append [ "elm.json", "src/Main.elm" ] (Job.inputFiles app)
```

...is that it lets us show in a dependency visualization that this job actually depends on the other job!